### PR TITLE
Merge `dev` to `feat-fast-sync` 28.01.2022

### PR DESCRIPTION
**Pull request by rafal-ch**
This PR integrates the recent changes from the `dev` (as of 28.01.2022 10:25am CEST) branch into the `feat-fast-sync` branch.

### CODE REVIEW

**Review by Fraser999 — APPROVED** [review]

**Review by nilz3ro — APPROVED** [review]

**Comment by rafal-ch** [comment]
Do not merge, there was an issue with a test on the previous build. I'd like to double check before merging:
https://drone-auto-casper-network.casperlabs.io/casper-network/casper-node/3960/3/3

**Comment by rafal-ch** [comment]
Merging, as the uncovered test flakiness is not related to this PR and is addressed in different PR: https://github.com/casper-network/casper-node/pull/2633

**Pull request merged**
